File: src/domain/entities/catalogs.ts
// Catálogos espejo de src-tauri/src/domain/catalogs.rs (REQ-05-01).
//
// HALLAZGO SERDE (verificado empíricamente con cargo en el ciclo F5): los
// enums unit serializan con el NOMBRE de la variante Rust tal cual
// ("Salario", "CuotasDeuda", "Avalanche"), no con la clave canónica
// minúscula de `as_str()`. Las uniones TS usan esos valores de cable;
// las claves canónicas del backend quedan expuestas como catálogo derivado.

export const INCOME_SOURCES = ['Salario', 'Freelance', 'Arriendos', 'Otros'] as const;

/** Fuente de ingreso mensual (valor tal cual llega por IPC). */
export type IncomeSource = typeof INCOME_SOURCES[number];

/** Claves canónicas del backend (`IncomeSource::as_str`), mismo orden. */
export const CANONICAL_INCOME_KEYS = [
  'salario',
  'freelance',
  'arriendos',
  'otros',
] as const;

export const INCOME_SOURCE_LABELS: Record<IncomeSource, string> = {
  Salario: 'Salario',
  Freelance: 'Freelance',
  Arriendos: 'Arriendos',
  Otros: 'Otros',
};

export const EXPENSE_CATEGORIES = [
  'Vivienda',
  'Alimentacion',
  'Transporte',
  'CuotasDeuda',
  'Ocio',
  'Otros',
] as const;

/** Categoría de gasto mensual (valor tal cual llega por IPC). */
export type ExpenseCategory = typeof EXPENSE_CATEGORIES[number];

/** Claves canónicas del backend (`ExpenseCategory::as_str`), mismo orden. */
export const CANONICAL_EXPENSE_KEYS = [
  'vivienda',
  'alimentacion',
  'transporte',
  'cuotas_deuda',
  'ocio',
  'otros',
] as const;

export const EXPENSE_CATEGORY_LABELS: Record<ExpenseCategory, string> = {
  Vivienda: 'Vivienda',
  Alimentacion: 'Alimentación',
  Transporte: 'Transporte',
  CuotasDeuda: 'Cuotas de deuda',
  Ocio: 'Ocio',
  Otros: 'Otros',
};

export const INVESTMENT_FAMILIES = ['RentaFija', 'RentaVariable', 'FincaRaiz'] as const;

/** Familia de inversión (valor tal cual llega por IPC). */
export type InvestmentFamily = typeof INVESTMENT_FAMILIES[number];

export const INVESTMENT_FAMILY_LABELS: Record<InvestmentFamily, string> = {
  RentaFija: 'Renta fija',
  RentaVariable: 'Renta variable',
  FincaRaiz: 'Finca raíz',
};

export const DEBT_STRATEGIES = ['Avalanche', 'Snowball'] as const;

/** Estrategia de ataque a la deuda (valor tal cual llega por IPC). */
export type DebtStrategy = typeof DEBT_STRATEGIES[number];

export const DEBT_STRATEGY_LABELS: Record<DebtStrategy, string> = {
  Avalanche: 'Avalancha',
  Snowball: 'Bola de nieve',
};

File: src-tauri/src/domain/catalogs.rs
//! Catálogos exactos del requerimiento (REQ-03-01/08): fuentes de ingreso
//! y categorías de gasto. Fuera de catálogo ⇒ error nombrado.

use serde::{Deserialize, Serialize};

use crate::domain::errors::{UnknownCategoryError, UnknownSourceError};

/// Fuente de ingreso mensual admitida.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum IncomeSource {
    Salario,
    Freelance,
    Arriendos,
    Otros,
}

impl IncomeSource {
    /// Catálogo exacto, en el orden del requerimiento.
    pub const ALL: [IncomeSource; 4] = [
        IncomeSource::Salario,
        IncomeSource::Freelance,
        IncomeSource::Arriendos,
        IncomeSource::Otros,
    ];

    /// Clave canónica: minúsculas sin tildes.
    pub fn as_str(self) -> &'static str {
        match self {
            IncomeSource::Salario => "salario",
            IncomeSource::Freelance => "freelance",
            IncomeSource::Arriendos => "arriendos",
            IncomeSource::Otros => "otros",
        }
    }

    /// Rechaza claves fuera de catálogo con error nombrado (REQ-03-08).
    pub fn parse(raw: &str) -> Result<Self, UnknownSourceError> {
        Self::ALL
            .iter()
            .find(|source| source.as_str() == raw)
            .copied()
            .ok_or_else(|| UnknownSourceError::new(raw))
    }
}

/// Categoría de gasto mensual admitida.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum ExpenseCategory {
    Vivienda,
    Alimentacion,
    Transporte,
    CuotasDeuda,
    Ocio,
    Otros,
}

impl ExpenseCategory {
    /// Catálogo exacto, en el orden del requerimiento.
    pub const ALL: [ExpenseCategory; 6] = [
        ExpenseCategory::Vivienda,
        ExpenseCategory::Alimentacion,
        ExpenseCategory::Transporte,
        ExpenseCategory::CuotasDeuda,
        ExpenseCategory::Ocio,
        ExpenseCategory::Otros,
    ];

    /// Clave canónica: minúsculas sin tildes.
    pub fn as_str(self) -> &'static str {
        match self {
            ExpenseCategory::Vivienda => "vivienda",
            ExpenseCategory::Alimentacion => "alimentacion",
            ExpenseCategory::Transporte => "transporte",
            ExpenseCategory::CuotasDeuda => "cuotas_deuda",
            ExpenseCategory::Ocio => "ocio",
            ExpenseCategory::Otros => "otros",
        }
    }

    /// Rechaza claves fuera de catálogo con error nombrado (REQ-03-08).
    pub fn parse(raw: &str) -> Result<Self, UnknownCategoryError> {
        Self::ALL
            .iter()
            .find(|category| category.as_str() == raw)
            .copied()
            .ok_or_else(|| UnknownCategoryError::new(raw))
    }
}
